src/lib.rs

```diff
@@ -1258,30 +1121,109 @@ pub enum Descriptor {
 #[repr(C)]
 pub struct File {
     fd: wasi_filesystem::Descriptor,
-    position: u64,
+    position: Cell<u64>,
 }
 
 const PAGE_SIZE: usize = 65536;
 
-impl Descriptor {
-    fn get(fd: Fd) -> &'static mut Descriptor {
+/// The maximum path length. WASI doesn't explicitly guarantee this, but all
+/// popular OS's have a `PATH_MAX` of at most 4096, so that's enough for this
+/// polyfill.
+const PATH_MAX: usize = 4096;
+
+struct State {
+    buffer_ptr: Cell<*mut u8>,
+    buffer_len: Cell<usize>,
+    ndescriptors: usize,
+    descriptors: MaybeUninit<[Descriptor; 128]>,
+    path_buf: MaybeUninit<[u8; PATH_MAX]>,
+}
+
+extern "C" {
+    fn get_global_ptr() -> usize;
+    fn set_global_ptr(a: usize);
```

Comment:
In the documentation above, these use `*mut u8` instead of `usize`; would it be feasible to use those types here too? Maybe that depends on why `get_global_ptr` might return 1 below though.

Reply:
Ah good point using `*mut T` works here so I'll update to that.